src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+
+use std::vec::Vec;
+
+use openssl::bn::{BigNum, BigNumContext};
+use openssl::ec::{EcGroup, EcKey, EcPoint, PointConversionForm};
+use openssl::ecdsa::EcdsaSig;
+use openssl::error::ErrorStack;
+use openssl::hash::{DigestBytes, Hasher, MessageDigest};
+use openssl::nid::Nid;
+use openssl::pkey::{Id, PKey, Public};
+use openssl::rsa::Rsa;
+use openssl::sign::Verifier;
+
+use super::common::{rsa_exponent_modulus, AlgorithmError, DigestType};
+use crate::base::iana::SecAlg;
+use crate::rdata::Dnskey;
+
+//============ Error Types ===================================================
+
+//----------- FromBytesError -----------------------------------------------
+
+/// An error in importing a key pair from bytes into OpenSSL.
+#[derive(Clone, Debug)]
+pub enum FromBytesError {
+    /// The requested algorithm was not supported.
+    UnsupportedAlgorithm,
+
+    /// The key's parameters were invalid.
+    InvalidKey,
+
+    /// An implementation failure occurred.
+    ///
+    /// This includes memory allocation failures.
+    Implementation,
+}
+
+//--- Conversion
+
+impl From<ErrorStack> for FromBytesError {
+    fn from(_: ErrorStack) -> Self {
+        Self::Implementation
+    }
+}
+
+//--- Formatting
+
+impl fmt::Display for FromBytesError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Self::UnsupportedAlgorithm => "algorithm not supported",
+            Self::InvalidKey => "malformed or insecure private key",
+            Self::Implementation => "an internal error occurred",
+        })
+    }
+}
+
+//--- Error
+
+impl std::error::Error for FromBytesError {}
+
+//----------- GenerateError --------------------------------------------------
+
+/// An error in generating a key pair with OpenSSL.
+#[derive(Clone, Debug)]
+pub enum GenerateError {
+    /// The requested algorithm was not supported.
+    UnsupportedAlgorithm,
+
+    /// An implementation failure occurred.
+    ///
+    /// This includes memory allocation failures.
+    Implementation,
+}
+
+//--- Conversion
+
+impl From<ErrorStack> for GenerateError {
+    fn from(_: ErrorStack) -> Self {
+        Self::Implementation
+    }
+}
+
+//--- Formatting
+
+impl fmt::Display for GenerateError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Self::UnsupportedAlgorithm => "algorithm not supported",
+            Self::Implementation => "an internal error occurred",
+        })
+    }
+}
+
+//--- Error
+
+impl std::error::Error for GenerateError {}
+
+//----------- DigestContext --------------------------------------------------
+
+/// Context for computing a message digest.
+pub struct DigestContext(Hasher);
+
+impl DigestContext {
+    /// Create a new context for a specified digest type.
+    pub fn new(digest_type: DigestType) -> Self {
+        Self(
+            match digest_type {
+                DigestType::Sha1 => Hasher::new(MessageDigest::sha1()),
```

Comment:
I can't find any trace of `MessageDigest::sha1()` in the original code. Is it intentional to add support for it here in the new code?

Reply:
NSEC3 uses SHA1...